ssl/ssl_sess.c

```diff
@@ -1081,58 +1102,76 @@ static void SSL_SESSION_list_remove(SSL_CTX *ctx, SSL_SESSION *s)
     s->prev = s->next = NULL;
 }
 
+/* session cache is already locked by caller */
 static void SSL_SESSION_list_add(SSL_CTX *ctx, SSL_SESSION *s)
 {
-    if ((s->next != NULL) && (s->prev != NULL))
-        SSL_SESSION_list_remove(ctx, s);
+    SSL_SESSION_CACHE *session_cache = ctx->session_cache;
 
-    if (ctx->session_cache_head == NULL) {
-        ctx->session_cache_head = s;
-        ctx->session_cache_tail = s;
-        s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
-        s->next = (SSL_SESSION *)&(ctx->session_cache_tail);
+    if (session_cache == NULL)
+        return;
+
+    SSL_SESSION_list_remove(ctx, s);
+
+    if (session_cache->head == NULL) {
+        session_cache->head = s;
+        session_cache->tail = s;
+        s->prev = (SSL_SESSION *)&(session_cache->head);
+        s->next = (SSL_SESSION *)&(session_cache->tail);
     } else {
-        s->next = ctx->session_cache_head;
+        s->next = session_cache->head;
         s->next->prev = s;
-        s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
-        ctx->session_cache_head = s;
+        s->prev = (SSL_SESSION *)&(session_cache->head);
+        session_cache->head = s;
     }
 }
 
 void SSL_CTX_sess_set_new_cb(SSL_CTX *ctx,
                              int (*cb) (struct ssl_st *ssl, SSL_SESSION *sess))
 {
-    ctx->new_session_cb = cb;
+    CRYPTO_THREAD_write_lock(ctx->lock);
+    CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+    ctx->session_cache->new_session_cb = cb;
+    CRYPTO_THREAD_unlock(ctx->session_cache->lock);
+    CRYPTO_THREAD_unlock(ctx->lock);
 }
 
-int (*SSL_CTX_sess_get_new_cb(SSL_CTX *ctx)) (SSL *ssl, SSL_SESSION *sess) {
-    return ctx->new_session_cb;
+int (*SSL_CTX_sess_get_new_cb(SSL_CTX *ctx)) (SSL *ssl, SSL_SESSION *sess)
+{
+    return ctx->session_cache->new_session_cb;
 }
 
 void SSL_CTX_sess_set_remove_cb(SSL_CTX *ctx,
                                 void (*cb) (SSL_CTX *ctx, SSL_SESSION *sess))
 {
-    ctx->remove_session_cb = cb;
+    CRYPTO_THREAD_write_lock(ctx->lock);
+    CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+    ctx->session_cache->remove_session_cb = cb;
+    CRYPTO_THREAD_unlock(ctx->session_cache->lock);
+    CRYPTO_THREAD_unlock(ctx->lock);
 }
 
 void (*SSL_CTX_sess_get_remove_cb(SSL_CTX *ctx)) (SSL_CTX *ctx,
                                                   SSL_SESSION *sess) {
-    return ctx->remove_session_cb;
+    return ctx->session_cache->remove_session_cb;
```

Comment:
I see your point, but it only becomes a problem if the memory is unmapped when freed. The chance that this occurring is infinitely small, but non-zero.
But, the problem you are referring to is slightly bigger as the old session cache is freed before the new one is assigned, so there is an extremely small window where the old session_cache may be freed before the new one is assigned.
I have not had the chance to run it through thread-sanitizer.
But the bigger picture I'm interested in is if these functions belong with the SSL_SESSION_CACHE, or should stay with the SSL_CTX.